Using forced Negotiate auth for reused connections

DarovskikhAndrei opened the pull request:
Added forced negotiate auth for reused connections.

If curl library reuses connection 
- with picked negotiate auth 
- credentials are not changed 
- negotiate auth is completed in reused connection

then curl uses forced negotiate auth (sends token in first http request)

old:
![curl7 42](https://cloud.githubusercontent.com/assets/10544047/8207161/06ae0b96-1508-11e5-8b4a-0702f8408d02.png)
new:
![curl7 42_forcednegotiate](https://cloud.githubusercontent.com/assets/10544047/8207162/06afada2-1508-11e5-8781-c8bd7b06211e.png)

fiddler logs.
old:

```
GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/7.5
WWW-Authenticate: Negotiate
X-Powered-By: ASP.NET
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 1293
Proxy-Support: Session-Based-Authentication

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;} 
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;} 
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} 
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUDulaaY6TPn+vNH4ZqGJy0U3H/zubk4QM99N7X5O1ABakI7gyD8jrwatdk+ljqohyo80m29H154eVZwI9nnSvZaxwyGdij3tna4F99fPLsjmrGOvvvbGKFw5CqiKzwrUoatbKLnaG4F29AKv/8G4DAgKaR5B+qb0ToBOTx72HYQG9qhGrcW67f2OgzfBShB1gsPhyYOOVJx7xLrln+g73gv9nWE+DDMbK+0jq/ux5EPCJJ6ZaM8y7iHTP4W2vHNOmQXEbIUbwc8mDdJzU4ej56CRQAzXkPitF+LZLoyPrHKVqBvgDnPnNhcYhNQ3Nr6AL2EoueNl3ccH1l/3DL9jDn+LvWaKFMU+/wkFc0RGSX7Z7VuSy6VEnhXoksfE/OOL+mcl27uI2sgK2K9MZIcvXUNV05LpHHPMtudrxL8dwzP9A==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuz3+p43LZMJrUx2IZVEQFFbbLXqzOEGtIDU9FUyO482AFqS+l+KPYmGXa9zdtGIfiGDpJpZvg+K74PTiXMmv4Zfebfgep2/s1DNrqLX2Gc4rAHn+jcxucELXcooYfx0WF2amagaWGLTDFE/dAefs=
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/7.5
WWW-Authenticate: Negotiate
X-Powered-By: ASP.NET
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 1293
Proxy-Support: Session-Based-Authentication

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;} 
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;} 
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} 
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUCyvdPZEi7/M321lrgYfJU+teZ5ANXXIYq9yLl8+bRNqyTJk3o8ksbk743ttR/WSedH1/s3se2kMwOJW21piHGl/0J5O0ImrKCvR5+UCdO5N4SvefXgAT17AyWam1EFVGNS7ntWZ2iB4pxuKh6SQdMxQzOs4wR4+xi2eANaD59K01PC8dg9zFa0oAVDMYmoirtO9cEHgwIfd9vf08ZyG0nkwg+ajV3DeQl4NjRRY6Qs8R4sRsZSUR1dJef+Cc4FHHKWgQAyii48nhdhYaEMOPoFAwaBTKKpws3ljMm1SL7L4wQEFFoJf6dhiqitrUxehrbY7MVtSf/qWfts4oeILiPY1Eypm+/1j5Bl0h8l9azzhV4STv7GCxDcjC9OJbgRdY1YWqkh/SFJkSjo2cprRe3fZayQ/dB9glKc/MdhOdpBVA==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARu+EBlaJuQbNSf866TxfX72fFhN8KXfI6XXMM/OFG+JuqneNxGC/gq+K8GFYsQIpjAD8quH69pZHPrR3gd+8hRymO/75sqgo8xpbDHN44fr1muc3dX6CAyhFUb0yvHBDgqdtveme/uquobSpdr250=
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/7.5
WWW-Authenticate: Negotiate
X-Powered-By: ASP.NET
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 1293
Proxy-Support: Session-Based-Authentication

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;} 
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;} 
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} 
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUBUBedFC7uoTINrWpM9zFlA6qSBZ9gx95HpNTV4fGb2AqL1xArnI64+eyl/JTh9CsiUDrCHxFMHnKpsy8wihGFjhVU+d9D+OpLIWwfaMY693aVWOW9IHuPS4E+m9h2djvypK+7Xs1oZfAaWDtHf7gW/9LK5l+pJH+TGwfLrVBZ24yQduif/DsMcnzxznXQAEoxlAB6N9tNImDkbyWDTPwQZvuAX+NW6RuCI5S68kAitjQ4iGDj/EHiWHCIRVFO/W45hVNTIEB5aJCzvTiD4zZ+V9PFYjfnZqb8sHt0CcoOSS0uTbaTbxYW4Pti273iRJfMcQk7IpxTMmk+cTlsDaaJ2eF4Wok4E3UzxHwh820WEwN4WSZgZ32y46NZD8Q8VmiNMHjRxX7Y+3c78Kw48Hx4wYgMK47U3At87gArVqbLKeA==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuzxgf2WY6QDxjS/dHYVvFL1WJl6HrPBNf4FdO/B0rjArVTrTE8qiy4WSg4kD+V1pygoZGYVkt+JXSz8VgpVhoXA9ZiMyep7Ijk7KpyWWwPxwO08dPgMfA5XmRIP8+NaJEq97GQ7BDP6VeQ9UrWpE=
Date: Wed, 17 Jun 2015 08:30:42 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------
```

new:

```
GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/7.5
WWW-Authenticate: Negotiate
X-Powered-By: ASP.NET
Date: Wed, 17 Jun 2015 09:17:35 GMT
Content-Length: 1293
Proxy-Support: Session-Based-Authentication

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;} 
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;} 
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} 
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUDRt+VTUBXRCPpkgBSCiEWjzY0aboAzSqQXBEkwtz2Jhad2fjdLjFn/35EV00ExT4TYD2Ls/0D/QnSaHmnZ5gYARk/cOKCGYXE6tUKDQhfeIAFeWPzrXWkKsiiisfGqWrjVDeM3ZuhSr91pRtzNxmoP7zLt63GsYHDDKzyRJnhBdDcSCva7xGpwWrewldwAUbcTn6dvIUxeOhwRbO4LpZoXZBISUV9dpznUR6HbdeV1+7TdAggdvbJjFn2qCLaMNJPSObsF1zTQojytiU8u7fGGlykh75L95k3zHHq8+mSeXs5UVEVxXKpCLFAxKQZWSXNqS2kQw4iCdAXzCcTdGPOiT6g8yzYr/hSmoyJZbrLnGMExCSgRqBT/ZEsg5bW2t//HVMBxUGynB39ukQDxLzTKU/afldbJWus805uUhrck/w==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuB80aviL/msIAMakXdGm5zqZRLTluP8/HrGSKDhvugCGx5xRbh4yEIlEXQGqSrgaJOgi0mzvWkfQx0RTzPzZL1TMx/w3p4py8D0iO777D+XUT9aOos2ClRfHlsrSs7zMvHZXiSK4ZCUz9IvFGw6o=
Date: Wed, 17 Jun 2015 09:17:35 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUBtS1gOc3ZjhL8pWskz0kGhl/00WiAnCUKwwbmF/G+gLYqfuVrOscpIsceEsZ28PE3Se1jVoCdiZGHCOOGSLQdpmWUrsjyJGF2CJuxI+LngCmJr7Ws5G3hWeTXGLPGpRpypJs/hVH6YhNU/h4Sy5DP4orkGA+OFVssc1t+e09xDI/5xyFLPhKdtPWoUF1gsROI/PwiNaCMCUuaEvMse6ICfpt6zyw4JFzA85rA0nDax0B+803psY5SCkY94sqlhLMbM/ic7zMnuKZ3RXw/L9jUroC1+cMphHFCKlPoNV62aXk63iRy163hqcPrXqJes0BqzItWB8M9pQhz1WZxZvDpHw7oCTCb0fQ5i2SR5GkFzKKBUYfhHXw5947oOU1xxfsfSNTRMKSg49GRxCE1eR58ti+lJuUUDLuf6UUEd8U2TsQ==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuI7MwGbHvLZPOaH/t8bg5fzB6fcrVSqFiRa+WGj0s+NaM9Ih1BNeuoUxYgGAH1WQaYgPmohqUI9+D3rinFqAX3BaWXBT3pslTYhhEl09iiVuufYwq/1GNMGApc2Fhuq6r71q1Xkw5Uy3rfymPqdI=
Date: Wed, 17 Jun 2015 09:17:35 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------

GET http://er57-srv/testNegotiate/index.html HTTP/1.1
Host: er57-srv
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogYABgAyAAAACgAKACgAAAAGAbEdAAAAD0RBUk9WU0tJSEFERVBUMDc=
Authorization: Negotiate YIIHKQYGKwYBBQUCoIIHHTCCBxmgMDAuBgkqhkiC9xIBAgIGCSqGSIb3EgECAgYKKwYBBAGCNwICHgYKKwYBBAGCNwICCqKCBuMEggbfYIIG2wYJKoZIhvcSAQICAQBuggbKMIIGxqADAgEFoQMCAQ6iBwMFACAAAACjggVaYYIFVjCCBVKgAwIBBaEIGwZERVBUMDeiGzAZoAMCAQKhEjAQGwRIVFRQGwhlcjU3LXNydqOCBSIwggUeoAMCARehAwIBB6KCBRAEggUMNNFe6nmN0GXN6vW5ZEvqCQo1/jgBQjID9u100Eo+WX8NIeNO6W+TM3kB1sbV9frCkltiSV3z8W8V3aV2oeU3NDhni+7kdFYtoXTGrp5WX5I7wt61l36/ItNQ4KWdL8Yn7nzj8iHiEE6rlQFq96jOy9QsxQlgRV2UZ8bdBqQSopsHMf7n6vJFzGj4kOK5XwRw3j9pBB7HzDYRgu/nzP7rbflGJPms1bfpulI/hyDp1Sw8wkTnWwlsaYQGvt4FXfjYq+VmG1Z7f/oRvVoV5PESzCpFpfITb063oneSyE7efGXuSXn08rf5A4J/MpcSsw7jKJGHVB7usova8o2IvE8P/3lljn+ffaOKOM9xlXoAqvTDbw/zryzeymdKhsRr+TgNI1KHwciyZZgEZIeMF1Qja6FPVGaaR8VnLY25RmW+aYBgFuSUdqLrn+Qd5sdWXtecVKHV8YEZG4TZmdhBx3ltq8yDd8n4ZzCgDZYNPbuQYtx2p47p+t8lr47apJVtouANaFG0P1VsHF8Fsnk/d7HfKVUu9wiAYA5LEtkV6SEjFM4idp6nLpmd0j59p9ISRM4ViRoxCgYOQu73iK4RfXPfa11i3Nu6KoQs8ijZkV7c+L6O4Q7sUl1Hlm69jgtjlzV1cQv8dWMtQcOTbh3jsHAV7u7sBSMhu9JXyZ+lUuLuroSsnP0iY6EUCtr94aQ991ew1ur9QoaWqqFqbSdioZD906T7AT3P0QBt1LVQneMl0uUz6N0Hqe/NzocKnZ69qS1LCCZF9Lwg5Uj4duzenWfqu/pDmJhVf0MeAtx0IFAX2DehM6bPY99Fkst0fpdEogbqf9FEz3uvBNaCw4j6Rs7xJ5MwMzYoRv7A1WHe0Gdv60ntzG3AjrBpTGAcQ+5y4GpJKyWuHebU1AX8PTZU0fe0vtPjZ7dxEasPd+jiaJ9hMSemWbFWSSKQa41RhpElMT2oZqhbxI1EIhErpIyTLG58l1UBaIjLm965VnnCGb8+eR6pxYH6WWgSVO+cmezAWZ8X/lW+gT/7YCt9n53Mj2GEJeyFDf/xCxIlwkf3kP3JuZj9lIUmlwlmo7n1nss9aOurvvLDWV6aBIa8IB1ir2g5Icux5KRxl5Q7ZDa5RJV2zde52/xufCW3B6Y6KFUbVhf0sdc5UP0hrL3bbwRn4RWn1k7u8gUKiDq/zo+xzCY1c8AzNy2jCR1zqOEPGPkCGCZ1bkT79OEvoKiQGyMMz0w/I0zbVa7sPHAUH20W8RpCN13amSxfR3ofk9wbDofteE+vV7M1R7Yrsv3ZHULy4BRB/06lCJwacJHeKFX0owLO8a1qtpCnX7fJNgFS4j90Cmcvi/H1EZQiUpJslfIUMNVVCEZi8YS1VH9RC+mvKyAftEAnZ2WYJVMg/XrhBPdLp/40a/+eLUGuIJIqdCHzqJToPkuOH2sc1sYLwanaoe5ZLYNwMz4mg5CbJ2NQIGUcyYYgF80Ccx8ZMkkoage7b5L/nn3bchGXtp1wwqlXlqhtB0Qt//MGhxU1gncBCWyUCDYrOUSZmn012haETIKG6YaoXsX9WLDfNTbfD1WW+pMDKAAXY80fR+q0IOYR64vMcs4PO7UxC9et3WBOXGK6IxbkCJREOJJdomREWprCV1PP98H+HuCb6egYVELNITh+DQQWspmgW9wkgJJNMRINzsj97df+MyHPoKganW7FscD2QltXnD8A1X1qWyMIOtOkggFRMIIBTaADAgESooIBRASCAUAkx7d53cmjlYwMIpdP8bS7DbfwG9QwWK/4B9oRReqRwwkBVOXCDauSRhK09sQZEEOrLA2UR/AkWsIBmX6sqv6zHWaUdC3MH/KujiW67rpdvmpbgb2kbuEAx+X6N6iyyXRFZ+K6Q4xx/2RJtQJJ2h7P+VmHzyd2kQI9og/Hf1SJ0cSHCi2BE9Wh8HXSLEpLPOR5QFBLKUINaxL/CglB03y5JLOUOyoRT/bINWztkgqxsR5CAmVGXDnG3zoiGSOzXQbiBbgB9gQRlxzhLCIU6s34lOQcVp0rF9nkrg1C7Hhjr4oaeHdjCZjqmyqcJu6fEg7W4PIWBFGhSsfuJb7gtViwwhkk/BsItVRjYHoC6gQi6Z9rw3+YtqSfTta2LcTM5a69ne11UZ4mvMyQG2YeLEmU1rlGQ4Cp1DABbYWCGkDNUg==
User-Agent: curl/7.43.0-DEV
Accept: */*
Connection: Keep-Alive


HTTP/1.1 200 OK
Content-Type: text/html
Last-Modified: Wed, 12 Nov 2014 11:02:01 GMT
Accept-Ranges: bytes
ETag: "1150d21568fecf1:0"
Server: Microsoft-IIS/7.5
Persistent-Auth: false
X-Powered-By: ASP.NET
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuRQX7x/Avgu3etREDsDuctiPFMoAq6UmZFRC8z0p5K+S40Twe0PN/oMTCI3KQlLGLL6I/ojODjaynBG9LUWfFblOD7/w+zCg2b+pBZa6nlrIJG+y24zEOi2EXfLPl2enDP83VZe51OqugbZooHhU=
Date: Wed, 17 Jun 2015 09:17:35 GMT
Content-Length: 98

<!DOCTYPE html>
<html lang="ru" dir="ltr" class="redesign no-js">
<body>
hello world
</body>


------------------------------------------------------------------
```

frenche commented:
I didn't got thoroughly thru the code but please keep in mind that Negotiate may use system credentials so you have not user/pass to compare (the ones provided by the app are not used).

Arguably in this case it is rare where the system credentials change but some server side application are doing this (by changing environment variable settings).

DarovskikhAndrei commented:
>Arguably in this case it is rare where the system credentials change but some server side application are doing this (by changing environment variable settings).

For all requests curl library send new actual token. Credentials comparing only for choose connection.
May be add option to rule it?

frenche commented:
Sorry, I don't fully understand, I'll try look further at the code.

DarovskikhAndrei commented:
changes by files:

urldata.h
- added new auth state GSS_AUTHCOMPLETE. it state using for correct finished connections with picked negotiate auth.

http.c: 
- added setting GSS_AUTHCOMPLETE for connections after correct executed request. 
- not change state before request if current state is GSS_AUTHCOMPLETE.

url.c: 
- modified function for select exists connections to support new conditions for forced negotiate auth (check credentials and reset GSS_AUTHCOMPLETE state if need).

http_negotiate_sspi.c:
- moved function cleanup.
- added function for generate token and filling negotiate context
- added function for check possible using forced negotiate auth
- modified function output_auth_headers for support forced negotiate auth.

frenche commented:
I went roughly over the code and I'd like to raise a few points.

- I think it is more secure and better way to inquire the name of the credentials used and then compare them when trying to re-use a connection (similar to what we do in NTLM) rather than force new authorization header on the authenticated connection and hope the server will reset auth.

- When NTLM is used inside negotiate the connections should be reused so there is a need to query the context for what mechanism was used (in addition to the name).
Also in NTLM need to be careful not to restart at every request as it spans over more than one req.

- The suggested changes don't seem to address the GSSAPI counterpart of Negotiate code (only windows SSPI).

I've tried the approach I've suggested (querying the credentials and the context) in GSSAPI with some success (although I still have some unresolved items).
I am not sure SSPI provides the necessary interface but I think it does (QueryCredentialsAttributes, QueryCredentialsAttributes, QuerySecurityPackageInfo, etc). 

Regards.

p1ng0o commented:
seem #1095 linked
